CI: Update Elixir & OTP build matrix, support old OTP versions

sgerrand opened the pull request:
💁 These changes update the the matrix of versions used in continuous integration to include:
- Elixir 1.16, 1.17 and 1.18
- OTP 27

I've followed the existing pattern of incrementing the Elixir minor version in lockstep with the most recent major version of Erlang/OTP that's supported by that Elixir version.

⚠️ The other significant part of these changes is to mitigate GitHub's recent retirement of `ubuntu-20.04` as a runner image which can't be used by a GitHub Actions workflow job any more.
```
This is a scheduled Ubuntu 20.04 retirement. Ubuntu 20.04 LTS runner will be removed on 2025-04-15. For more details, see https://github.com/actions/runner-images/issues/11101
```
To minimise the size of changes I've split the original test job into 2 separate jobs:
- the original one, without OTP <= v23.
- a new one with a matrix of those older OTP versions that runs inside a [dedicated Ubuntu 20.04 container](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer).

sgerrand committed:
Test against Elixir 1.16 and Erlang/OTP 27 in CI

sgerrand commented:
👋 @michalmuskala, would you please review this change when you have a moment.

michalmuskala commented:
Sorry for the delay. Looks like there's an issue with the chosen OTP/Elixir version pair.

In general the pattern we tried to follow is to test every dot Elixir release with the latest Erlang it supports - if we end up with some Erlang version not covered we bump down one of the previous Elixir releases. That said, it gets out of sync. Feel free to make bigger changes to this to cover all the missing Elixir releases.

sgerrand commented:
No worries and thank you for the guidance. I'll take another approach to the work that I've done to date.

sgerrand committed:
Use OTP v26 with Elixir 1.16.x

See: https://hexdocs.pm/elixir/compatibility-and-deprecations.html#between-elixir-and-erlang-otp

sgerrand committed:
Test against Elixir 1.17 and Erlang/OTP 27

sgerrand committed:
Configure GNU compiler version per Ubuntu release

Recent versions of Ubuntu have different versions of gcc and g++
available, which neccesitates configuring versions per OS.

See GitHub's release notes for their runner images:
- https://github.com/actions/runner-images/blob/ubuntu22/20250427.1/images/ubuntu/Ubuntu2204-Readme.md
- https://github.com/actions/runner-images/blob/ubuntu24/20250427.1/images/ubuntu/Ubuntu2404-Readme.md

sgerrand commented on .github/workflows/ci.yml:
```diff
@@ -79,12 +115,12 @@ jobs:
       - run: mix compile --warnings-as-errors
         if: matrix.warnings_as_errors
         env:
-          CC: gcc-10
-          CXX: g++-10
+          CC: ${{matrix.gcc}}
+          CXX: ${{matrix.gpp}}
       - run: mix test
         env:
-          CC: gcc-10
-          CXX: g++-10
+          CC: ${{matrix.gcc}}
+          CXX: ${{matrix.gpp}}
```
The available versions of gcc and g++ change in each Ubuntu release. This change to the build matrix should make the correct executable available per OS version.
```
==> jason_native
 CXX jason.o
make: g++-10: No such file or directory
make: *** [Makefile:64: /home/runner/work/jason/jason/_build/test/lib/jason_native/obj/jason.o] Error 127
could not compile dependency :jason_native, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile jason_native --force", update it with "mix deps.update jason_native" or clean it with "mix deps.clean jason_native"
==> jason
** (Mix) Could not compile with "make" (exit status: 2).
You need to have gcc and make installed. If you are using
Ubuntu or any other Debian-based system, install the packages
"build-essential". Also install "erlang-dev" package if not
included in your Erlang/OTP version. If you're on Fedora, run
"dnf group install 'Development Tools'".
```

sgerrand committed:
Test against Elixir 1.18 and Erlang/OTP 27

sgerrand commented on .github/workflows/ci.yml:
```diff
@@ -18,42 +18,83 @@ jobs:
           - elixir: 1.4.x
             otp: 20
             os: ubuntu-20.04
```
This runner image has been retired and can't be used by a GitHub Actions workflow job any more.
```
This is a scheduled Ubuntu 20.04 retirement. Ubuntu 20.04 LTS runner will be removed on 2025-04-15. For more details, see https://github.com/actions/runner-images/issues/11101
```
The implications of this are that any Elixir & OTP pairs which used `ubuntu-20.04` as the runner image will break as the [`erlef/setup-beam` action](https://github.com/erlef/setup-beam) uses [Hex.pm's OTP builds](https://github.com/hexpm/bob/blob/main/README.md#erlang-builds) as the source of available builds. This means that OTP versions 20.x - 24.1 inclusive aren't available for download and installation.

I'll see if I can find a replacement image or another approach which enables keeping these older OTP versions in the build matrix.

sgerrand replied:
No luck in finding an immediate replacement for this deprecated image. However, I have successfully tested that OTP versions built for Ubuntu 20.04 work fine on Ubuntu 22.04. I plan to propose that Hex.pm's builder be extended to account for these deprecations of GitHub's `ubuntu-20.04` runner image.

sgerrand replied:
I ended up resolving this problem by creating a similar job in [71212eb](https://github.com/michalmuskala/jason/pull/200/commits/71212ebffed14493b3ec662ef6ef82666a952bec) and slicing the parts of the build matrix which depend on Ubuntu 20.04 in a container instead. It will make the workflow definition a little more cumbersome to maintain but the older Elixir & OTP pairs are unlikely to change.

sgerrand reviewed (commented):
⚠️ Found a breaking change, will try to resolve it.

sgerrand committed:
Run OTP versions which rely on ancient OpenSSL versions in container

GitHub retired the `ubuntu-20.04` runner image last month. These changes
add a new job which runs the Elixir & OTP pairs which need that Ubuntu
version in a container which uses the official Docker image instead.

A sample error from the existing GitHub workflow job follows:
```
This is a scheduled Ubuntu 20.04 retirement. Ubuntu 20.04 LTS runner
will be removed on 2025-04-15. For more details, see
actions/runner-images#11101
```

sgerrand commented:
👋 @michalmuskala: These changes are ready for your review again and should be mergable.

sgerrand commented:
👋 @michalmuskala, would you please review this change when you have a moment.

sgerrand commented:
👋 @michalmuskala, a gentle nudge to review these changes.

sgerrand commented:
🛎️ Ping?
